builder/dockerfile/dispatchers.go

```diff
@@ -199,53 +199,63 @@ func dispatchCopy(b *Builder, args []string, attributes map[string]bool, origina
 	return b.runContextCommand(args, false, false, "COPY", im)
 }
 
-// FROM imagename
-//
-// This sets the image the dockerfile will build on top of.
-//
-func from(b *Builder, args []string, attributes map[string]bool, original string) error {
+func parseContextName(args []string) (string, error) {
 	ctxName := ""
-	if len(args) == 3 && strings.EqualFold(args[1], "as") {
+	switch {
+	case len(args) == 3 && strings.EqualFold(args[1], "as"):
 		ctxName = strings.ToLower(args[2])
 		if ok, _ := regexp.MatchString("^[a-z][a-z0-9-_\\.]*$", ctxName); !ok {
-			return errors.Errorf("invalid name for build stage: %q, name can't start with a number or contain symbols", ctxName)
+			return "", errors.Errorf("invalid name for build stage: %q, name can't start with a number or contain symbols", ctxName)
 		}
-	} else if len(args) != 1 {
-		return errExactlyOneArgument("FROM")
+	case len(args) != 1:
+		return "", errExactlyOneArgument("FROM")
 	}
 
-	if err := b.flags.Parse(); err != nil {
-		return err
+	return ctxName, nil
+}
+
+func getFromImage(b *Builder, name string) (builder.Image, error) {
+	if im, ok := b.imageContexts.byName[name]; ok {
+		if len(im.ImageID()) > 0 {
+			return im, nil
+		}
+		// TODO: what is this case?
+		return nil, nil
```

Comment:
I don't think I understand this case. When would the imageContexts have a named context without an image ID?